File: Pine/SocketMessages/HelloMessage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace Pine.SocketMessages
{
	public class HelloMessage : Message
	{
		public static readonly UInt64 CURRENT_VERSION = 0x0;

		public HelloMessage()
		{
			Header.Type = MessageType.Hello;
			Header.BodySize = Size;
		}

		public override bool ParseBody(byte[] buffer)
		{
			if ((UInt64)buffer.LongLength != Size)
				return false;

			Version = BitConverter.ToUInt64(buffer, 0);
			return true;
		}

		public override byte[] Serialize()
		{
			byte[] buffer = Array.Empty<byte>();
			buffer = buffer.Concat(Header.Serialize()).ToArray();

			buffer = buffer.Concat(BitConverter.GetBytes(Version)).ToArray();

			return buffer;
		}

		public static readonly UInt64 Size = sizeof(UInt64);
		public UInt64 Version { get; set; } = 0;
	}
}

File: shared/src/socket_messages/hello_message.cpp
#include "socket_messages/hello_message.h"

#include <algorithm>
#include <bit>
#include <cstdint>
#include <vector>

#include "message.h"

namespace pine::socket_messages
{
	hello_message::hello_message()
	{
		header.type = message_type::HELLO_MESSAGE;
		header.body_size = get_body_size();
	}

	bool hello_message::parse_body(std::vector<uint8_t> const& buffer)
	{
		if (buffer.size() != get_body_size())
			return false;

		std::memcpy(std::bit_cast<void*>(&version), std::bit_cast<void*>(buffer.data()), sizeof(version));

		return true;
	}

	std::vector<uint8_t> hello_message::serialize() const
	{
		std::vector<uint8_t> buffer(message_header::size + get_body_size(), 0);
		std::vector<uint8_t>::iterator it = buffer.begin();

		it = std::copy_n(header.serialize().begin(), message_header::size, it);
		it = std::copy_n(std::bit_cast<uint8_t const*>(&version), sizeof(version), it);

		return buffer;
	}

	uint64_t hello_message::get_body_size() const
	{
		return sizeof(version);
	}
}